Remove old model usage from MailPoet\Cron\Workers\SendingQueue\Tasks\Newsletter - Part 1 [MAILPOET-4363]

rodrigoprimo opened the pull request:
## Description
Creating this PR as a draft while #4330 is not merged.

I'm opting to split [MAILPOET-4363] into multiple PRs to reduce the size of the code changes. More PRs to come once this one is merged.

## QA notes
Please perform exploratory testing around sending newsletters

## Linked tickets
[MAILPOET-4363]

[MAILPOET-4363]: https://mailpoet.atlassian.net/browse/MAILPOET-4363?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAILPOET-4363]: https://mailpoet.atlassian.net/browse/MAILPOET-4363?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo committed:
Replace most of the uses of Newsletter model with Doctrine in NewsletterTest

[MAILPOET-4363]

rodrigoprimo committed:
Replace SendingQueue model with Doctrine in NewsletterTest

[MAILPOET-4363]

rodrigoprimo committed:
Replace NewsletterSegment model with Doctrine in NewsletterTest

[MAILPOET-4363]

rodrigoprimo committed:
Replace Subscriber model with Doctrine in NewsletterTest

[MAILPOET-4363]

rodrigoprimo committed:
Refactor markNewsletterAsSent() to use Doctrine instead of Paris

[MAILPOET-4363]

rodrigoprimo committed:
Stop using a mock of NewslettersRepository to prevent errors in tests

This commit changes the SendingQueueTest test class to always use a real
NewslettersRepository object instead of a mock object when instantiating
SendingQueue. This is necessary to prevent false errors in the tests.

Using a mock of the NewslettersRepository in this test class means that
all code inside SendingQueue that uses this repository could behave in
unpredictable ways. As it was happening with a few tests incorrectly
failing after dd6e02e2c16a3a0ef7ad2c12479917e33c312c77:

https://app.circleci.com/pipelines/github/mailpoet/mailpoet/11132/workflows/7b4df151-5a10-48f0-9711-dd590a764880/jobs/192389/tests#failed-test-0

In this case, the code was not changing the status of the newsletter
because findOneById() was returning an empty Newsletter entity instead
of the real Newsletter entity that corresponds to the object in the
database. And, thus, $newsletter->getType() was returning `null` instead
of `standard`.

This problem was already partially addressed in the commit below that
removed some of the usages of the NewslettersRepository mock:

https://github.com/mailpoet/mailpoet/commit/5042cdd15e341b18d034619b7a2a7c762af7f6b9

[MAILPOET-4363]

rodrigoprimo committed:
Refactor Newsletter to use Doctrine instead of the NewsletterSegment model

[MAILPOET-4363]

rodrigoprimo committed:
Replace Subscriber model with Doctrine in Tasks\Newsletter

[MAILPOET-4363]

rodrigoprimo committed:
Add return type declaration for Newsletter::prepareNewsletterForSending()

[MAILPOET-4363]

rodrigoprimo committed:
Refactor Newsletter to use Doctrine instead of the Newsletter model

[MAILPOET-4363]

rodrigoprimo committed:
Use NewsletterEntity instead of Newsletter model in Links

[MAILPOET-4363]

rodrigoprimo committed:
Rename NewsletterTest::queue to NewsletterTest::sendingTask

Doing this for consistency and to better makr that this property
contains an instance of \MailPoet\Tasks\Sending and not an instance
of \MailPoet\Models\SendingQueue.

[MAILPOET-4363]

rodrigoprimo committed:
Set the type of all the Newsletter parameters that use the Sending class

[MAILPOET-4363]

rodrigoprimo committed:
Rename $queue to $sendingTask

Doing this for consistency and to make it more clear that $sendingTask
contains an instance of \MailPoet\Tasks\Sending and not an instance of
\MailPoet\Models\SendingQueue.

[MAILPOET-4363]

rodrigoprimo committed:
Refactor getNewsletterFromQueue() to use Doctrine instead of Paris

[MAILPOET-4363]

rodrigoprimo committed:
Replace some of the uses of SendingQueue model with Doctrine

[MAILPOET-4363]

lysyjan commented on mailpoet/tests/integration/Cron/Workers/SendingQueue/Tasks/NewsletterTest.php:
```diff
@@ -243,38 +249,46 @@ public function testItSavesNewsletterPosts() {
-    expect($updatedNewsletter->status)->equals(Newsletter::STATUS_SENT);
-    expect($updatedNewsletter->sentAt)->equals($queue->processedAt);
+    $updatedNewsletter = $this->newslettersRepository->findOneById($newsletter->getId());
+    assert($updatedNewsletter instanceof NewsletterEntity);
```
I thought that we wanted to avoid using the `assert` function in our code base. It should be possible here by using `$this->assertInstanceOf`. There are more places with this. I won't comment on all of them.

rodrigoprimo replied:
I failed to notice that some of the code that I moved around in this PR already contained calls to `assert()`. Thanks for pointing that out. I just added a new commit to remove all calls to `assert()` inside the integration tests that are modified in this PR. I left a call to `assert()` inside `SendingQueue.php` as it is not that trivial to replace and it will eventually get removed as we continue working on replacing Paris with Doctrine.

lysyjan commented on mailpoet/lib/Cron/Workers/SendingQueue/Tasks/Newsletter.php:
```diff
@@ -283,11 +289,20 @@ public function markNewsletterAsSent(NewsletterEntity $newsletter, $queue) {
-      ->findArray();
-    return Helpers::flattenArray($segments);
+  public function getNewsletterSegments(NewsletterEntity $newsletter) {
+    $newsletterSegments = $this->newsletterSegmentRepository->findBy(['newsletter' => $newsletter]);
```
NewsletterEntity contains a method `getSegmentIds`. Do you think that it would be possible to use it here and simplify our code?

rodrigoprimo replied:
Good point. I missed that. I will create a new commit removing this method and using NewsletterEntity::getSegmentIds().

lysyjan commented on mailpoet/lib/Tasks/Sending.php:
```diff
@@ -190,6 +193,13 @@ public function queue() {
     return $this->queue;
   }
 
+  public function getSendingQueueEntity(): SendingQueueEntity {
```
I guess this method may return null. But I'm not sure if we should solve it because it can be an edge case.

rodrigoprimo replied:
I'm not sure as well, so if you are ok with that, I will leave the code as is. We will probably need to revisit this method when we get to the ticket to refactor the Sending class.

lysyjan commented on mailpoet/lib/Cron/Workers/SendingQueue/SendingQueue.php:
```diff
@@ -272,12 +274,21 @@ public function processQueue($queue, $newsletter, $subscribers, $timer) {
         $this->newsletterTask->prepareNewsletterForSending(
-          $newsletter,
-          $subscriber,
+          $newsletterEntity,
```
I was wondering how this got through PHPStan because, if I am not wrong, the newsletter entity can be null here.
In the end, I found that the `newsletterTask` property is not annotated.

rodrigoprimo replied:
Good catch. I documented the properties and add an if to workaround the PHPStan error. The if is inside a foreach which is not really necessary but it was the easiest way that I found and it shouldn't be a big problem as far as I can see as we will revisit this soon as part of another Doctrine ticket.

lysyjan reviewed (commented):
I have a few minor remarks, but I'm okay with moving this PR forward. It's up to you if you want to look at it.

rodrigoprimo committed:
Replace calls to assert() with Assert::assertInstanceOf()

Doing this as it was mentioned in the review process of the PR.

For more context see: https://wp.me/pcNwfB-wR

[MAILPOET-4363]

rodrigoprimo committed:
Remove method Tasks\Newsletter\getNewsletterSegments()

This method was used only once inside SendingQueue and it can be
replaced by NewsletterEntity::getSegmentIds() to simplify the code.

[MAILPOET-4363]

rodrigoprimo committed:
Document the types of two SendingQueue properties that were undocumented

It was also necessary to add an if condition to check the type of a
variable before calling one of the methods of one of the documented
properties.

[MAILPOET-4363]

rodrigoprimo commented:
@lysyjan, I addressed all of your comments. When you have a moment, could you please check this PR again?

(the CircleCI build failed but due to a flaky test that is likely unrelated to the changes suggested here, I'm rerunning it and will rerun again if it fails one more time)

lysyjan reviewed (approved):
Thank you for your answers and for addressing them!

Aschepikov reviewed (changes requested):
Hi. During the testing I found one issue:
I can resend the newsletter multiple times to the same subscribers. And I don't see a warning that this newsletter was sent. Newsletters from other versions cant be resent. Newsletters from this version can be resent in others. On the main page of emails, I can see that it sent 6 emails. 
![image](https://user-images.githubusercontent.com/113185437/191266828-b21f9b3f-953c-40f0-8e26-f6ea9d5771cc.png)
But in the statistics, I can see that it was sent 12 times to 6 subscribers and I can edit and resend it for the third time
![image](https://user-images.githubusercontent.com/113185437/191267026-a12255c1-a91d-4367-b469-b1a9131007f0.png)

rodrigoprimo commented:
Hi, @Aschepikov. Thanks for testing this PR. I was not able to reproduce the problem that you mentioned. See the screencast below. Am I missing some steps to reproduce the issue? Thanks.

https://user-images.githubusercontent.com/77215/191339108-5c3a069a-2347-4d96-acb2-928d098c14f0.mp4

Aschepikov reviewed (approved):
Thank you, yes I can't reproduce it too, I have this broken newsletter and nothing else so it might be some edge case also present on the production, or even a server-side problem that has caused it.. and I will recheck it on the release build testing on Tuesday
 and I’ll be monitoring it in order to catch it once again and report it.
